src/chunkserver/scan_manager.cpp

```diff
@@ -0,0 +1,174 @@
+/*
+ *  Copyright (c) 2020 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * Created Date: April  13th 2021
+ * Author: huyao
+ */
+
+#include "src/chunkserver/scan_manager.h"
+
+namespace curve {
+namespace chunkserver {
+
+ScanManager::~ScanManager() {
+    return;
+}
+
+int ScanManager::Init(const ScanManagerOptions options) {
+    toStop_.store(false, std::memory_order_release);
+    scanSize = options.scanSize;
+    scanThread_ = Thread(&ScanManager::Run, this);  
+    waitInterval_.Init(options.intervalSec * 1000);
+    return 0;
+}
+
+int ScanManager::Fini() {
+    waitInterval_.StopWait();
+    return 0;
+}
+
+int ScanManager::Run() {
+    ScanKey key;
+    while (!toStop_.load(std::memory_order_acquire)) {
+        rwLock_.RDLock();
+        if (waitScanSet.empty()) {
+            rwLock_.Unlock();
+            waitInterval_.WaitForNextExcution();
+            continue;
+        }
+        rwLock_.Unlock();
+        key = Dequeue();
+        StartScanJob(key);
+    }
+
+    return 0;
+}
+
+void ScanManager::Enqueue(LogicPoolID poolId, CopysetID id) {
+    WriteLockGuard writeGuard(rwLock_);
+    ScanKey key(poolId, id);
+    if (waitScanSet.find(key) == waitScanSet.end()) {
+        waitScanSet.insert(key);
+    }
+    return;
+}
+
+ScanKey ScanManager::Dequeue() {
+    ScanKey key;
+    ReadLockGuard writeGuard(rwLock_);
+    key = *(waitScanSet.begin());
```

Comment:
this may cause a Segment Fault is set is empty.

Reply:
agreed